File: src/lib/api/mods.ts
import type {
  ExtendedModInfo,
  KableInstallation,
  ModFilter,
  ModInfoKind,
  ModJarInfo,
  ModpackContext,
  ModpackPrepareResult,
  ModrinthVersion,
  ProviderKind,
} from "$lib";
import { invoke } from "@tauri-apps/api/core";

// Unified mod/modpack download/prepare API
export async function downloadOrPrepareMod(
  provider: ProviderKind,
  modId: string,
  versionId: string | null,
  installation: KableInstallation,
): Promise<ModpackPrepareResult> {
  return await invoke("download_or_prepare_mod", {
    provider,
    modId,
    versionId,
    installation,
  });
}

import type { ModpackSelection } from "$lib";

export async function applyModpackSelection(
  installation: KableInstallation,
  selection: ModpackSelection,
  context: ModpackContext,
): Promise<void> {
  return await invoke("apply_modpack_selection", {
    installation,
    selection,
    context,
  });
}

export async function getMods(
  provider: ProviderKind,
  offset: number,
): Promise<ModInfoKind[]> {
  console.log(
    `[ModsAPI] Calling get_mods with provider: ${provider}, offset: ${offset}`,
  );
  const result = (await invoke("get_mods", {
    provider,
    offset,
  })) as ModInfoKind[];
  console.log(
    `[ModsAPI] get_mods returned ${Array.isArray(result) ? result.length : "non-array"} results:`,
    result,
  );
  return result;
}

export async function downloadMod(
  provider: ProviderKind,
  modId: string,
  versionId: string | null,
  installation: KableInstallation,
): Promise<void> {
  return invoke("download_mod", { provider, modId, versionId, installation });
}

export async function getProjects(
  provider: ProviderKind,
  projectIds: string[],
): Promise<ModInfoKind[]> {
  console.log(
    `[ModsAPI] Calling get_projects with provider: ${provider}, projectIds:`,
    projectIds,
  );
  const result = (await invoke("get_projects", {
    provider,
    projectIds,
  })) as ModInfoKind[];
  console.log(`[ModsAPI] get_projects returned ${result.length} projects`);
  return result;
}

export async function getProjectVersions(
  provider: ProviderKind,
  projectId: string,
  loaders?: string[],
  gameVersions?: string[],
): Promise<ModrinthVersion[]> {
  console.log(
    `[ModsAPI] Calling get_project_versions with provider: ${provider}, projectId: ${projectId}, loaders:`,
    loaders,
    "gameVersions:",
    gameVersions,
  );
  const result = (await invoke("get_project_versions", {
    provider,
    projectId,
    loaders: loaders || null,
    gameVersions: gameVersions || null,
  })) as ModrinthVersion[];
  console.log(
    `[ModsAPI] get_project_versions returned ${result.length} versions`,
  );
  return result;
}

export async function setProviderFilter(
  provider: ProviderKind,
  installation: KableInstallation | null,
  filter: ModFilter | null,
): Promise<void> {
  console.log(
    `[ModsAPI] Calling set_provider_filter with provider: ${provider}, installation:`,
    installation,
    "filter:",
    filter,
  );
  const result = await invoke("set_provider_filter", {
    provider,
    installation,
    filter,
  });
  console.log(`[ModsAPI] set_provider_filter completed:`, result);
  return result as void;
}

export async function setProviderLimit(
  provider: ProviderKind,
  limit: number,
): Promise<void> {
  return invoke("set_provider_limit", { provider, limit });
}

export async function clearProviderCache(
  provider: ProviderKind,
): Promise<void> {
  return invoke("clear_provider_cache", { provider });
}

export async function purgeStaleProviderCache(
  provider: ProviderKind,
): Promise<void> {
  return invoke("purge_stale_provider_cache", { provider });
}

export async function getExtendedModInfo(
  modJarInfo: ModJarInfo,
): Promise<ExtendedModInfo> {
  return invoke("get_extended_mod_info", { modJarInfo });
}

export interface ModMetadata {
  project_id: string;
  file_name: string;
  version_number: string;
  modrinth_version_id?: string | null;
  download_time: string;
}

export interface ModpackSourceRecord {
  provider: ProviderKind;
  mod_id: string;
  version_id?: string | null;
  modpack_name?: string | null;
  modpack_version?: string | null;
  installed_at: string;
  managed_project_ids: string[];
}

export async function getModMetadata(
  installation: KableInstallation,
  jarFilename: string,
): Promise<ModMetadata> {
  return invoke("get_mod_metadata", { installation, jarFilename });
}

export async function getModpackSourceRecords(
  installation: KableInstallation,
): Promise<ModpackSourceRecord[]> {
  return invoke("get_modpack_source_records", { installation });
}

File: src-tauri/src/commands/mods.rs
#[tauri::command]
pub async fn download_or_prepare_mod(
    provider: crate::mods::ProviderKind,
    mod_id: String,
    version_id: Option<String>,
    installation: crate::installations::kable_profiles::KableInstallation,
) -> Result<serde_json::Value, String> {
    // Call backend logic to handle mod or modpack
    match crate::mods::download_or_prepare_mod(
        provider,
        &mod_id,
        version_id.as_deref(),
        &installation,
    )
    .await
    {
        Ok(crate::mods::DownloadOrPrepareResult::ModInstalled) => {
            Ok(serde_json::json!({"success": true}))
        }
        Ok(crate::mods::DownloadOrPrepareResult::Modpack { modpack, context }) => {
            Ok(serde_json::json!({"modpack": modpack, "context": context}))
        }
        Err(e) => Err(e),
    }
}
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

#[tauri::command]
pub async fn complete_modpack_download(
    instance_id: String,
    modpack_id: String,
    mods: crate::installations::mrpack::PackFileDetailedGroup,
    resourcepacks: crate::installations::mrpack::PackFileDetailedGroup,
    shaderpacks: crate::installations::mrpack::PackFileDetailedGroup,
    install_dir: String,
) -> Result<(), String> {
    crate::mods::complete_modpack_download(
        instance_id,
        modpack_id,
        mods,
        resourcepacks,
        shaderpacks,
        install_dir,
    )
    .await
}

#[tauri::command]
pub async fn prepare_modpack_diff(
    mrpack_path: String,
    installation_dir: String,
    subfolder: Option<String>,
) -> Result<crate::mods::ModpackDiffResult, String> {
    crate::mods::prepare_modpack_diff(mrpack_path, installation_dir, subfolder).await
}

#[tauri::command]
pub async fn apply_modpack_selection(
    installation: crate::installations::kable_profiles::KableInstallation,
    selection: ModpackSelection,
    context: ModpackContext,
) -> Result<(), String> {
    crate::mods::apply_modpack_selection(installation, selection, context).await
}

#[tauri::command]
pub async fn extract_mrpack_and_load_index(
    mrpack_path: String,
    temp_dir: String,
) -> Result<MrpackIndex, String> {
    let mrpack_path = PathBuf::from(mrpack_path);
    let temp_dir = PathBuf::from(temp_dir);
    crate::mrpack::extract_mrpack(&mrpack_path, &temp_dir)
        .map_err(|e| format!("Failed to extract mrpack: {e}"))?;
    let index = crate::mrpack::load_index(&temp_dir)
        .map_err(|e| format!("Failed to load modrinth.index.json: {e}"))?;
    Ok(index)
}

#[tauri::command]
pub async fn list_pack_files_from_index(index: MrpackIndex) -> Result<PackFileGroups, String> {
    Ok(crate::mrpack::list_pack_files(&index))
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiffResult {
    pub conflicts: Vec<PackFileInfo>,
    pub new_files: Vec<PackFileInfo>,
}

#[tauri::command]
pub async fn diff_pack_files_against_profile(
    pack_files: Vec<PackFileInfo>,
    profile_dir: String,
    subfolder: String,
) -> Result<DiffResult, String> {
    let profile_dir = PathBuf::from(profile_dir);
    let (conflicts, new_files) =
        crate::mrpack::diff_pack_files(&pack_files, &profile_dir, &subfolder)
            .map_err(|e| format!("Failed to diff pack files: {e}"))?;
    Ok(DiffResult {
        conflicts,
        new_files,
    })
}

#[tauri::command]
pub async fn copy_selected_pack_files_to_profile(
    extracted_dir: String,
    profile_dir: String,
    files: Vec<PackFileInfo>,
    subfolder: String,
    overwrite: bool,
) -> Result<(), String> {
    let extracted_dir = PathBuf::from(extracted_dir);
    let profile_dir = PathBuf::from(profile_dir);
    crate::mrpack::copy_selected_pack_files(
        &extracted_dir,
        &profile_dir,
        &files,
        &subfolder,
        overwrite,
    )
    .map_err(|e| format!("Failed to copy selected pack files: {e}"))
}

#[tauri::command]
pub async fn extract_overrides_to_profile(
    extracted_dir: String,
    profile_dir: String,
) -> Result<(), String> {
    let extracted_dir = PathBuf::from(extracted_dir);
    let profile_dir = PathBuf::from(profile_dir);
    crate::mrpack::extract_overrides_to_profile(&extracted_dir, &profile_dir)
        .map_err(|e| format!("Failed to extract overrides: {e}"))
}
use crate::installations::kable_profiles::KableInstallation;
use crate::mrpack::MrpackIndex;
use crate::{mods::*, ModJarInfo, PackFileGroups, PackFileInfo};

#[tauri::command]
pub async fn get_mods(provider: ProviderKind, offset: usize) -> Result<Vec<ModInfoKind>, String> {
    crate::mods::get_mods(provider, offset).await
}

#[tauri::command]
pub async fn get_mod_metadata(
    installation: KableInstallation,
    jar_filename: String,
) -> Result<ModMetadata, String> {
    crate::mods::get_mod_metadata(&installation, &jar_filename).await
}

#[tauri::command]
pub async fn download_mod(
    provider: ProviderKind,
    mod_id: String,
    version_id: Option<String>,
    installation: KableInstallation,
) -> Result<(), String> {
    crate::mods::download_mod(provider, &mod_id, version_id.as_deref(), &installation).await
}

#[tauri::command]
pub async fn get_projects(
    provider: ProviderKind,
    project_ids: Vec<String>,
) -> Result<Vec<ModInfoKind>, String> {
    match provider {
        ProviderKind::Modrinth => {
            let projects = modrinth::get_projects(project_ids).await?;
            Ok(projects.into_iter().map(ModInfoKind::Modrinth).collect())
        }
        ProviderKind::CurseForge => {
            Err("CurseForge bulk project fetching not yet implemented".to_string())
        }
    }
}

#[tauri::command]
pub async fn get_project_versions(
    provider: ProviderKind,
    project_id: String,
    loaders: Option<Vec<String>>,
    game_versions: Option<Vec<String>>,
) -> Result<Vec<modrinth::ModrinthVersion>, String> {
    match provider {
        ProviderKind::Modrinth => {
            modrinth::get_project_versions_filtered(&project_id, loaders, game_versions).await
        }
        ProviderKind::CurseForge => {
            Err("CurseForge version fetching not yet implemented".to_string())
        }
    }
}

#[tauri::command]
pub async fn set_provider_filter(
    provider: ProviderKind,
    installation: Option<KableInstallation>,
    filter: Option<ModFilter>,
) {
    crate::mods::set_provider_filter(provider, installation.as_ref(), filter);
}

#[tauri::command]
pub async fn set_provider_limit(provider: ProviderKind, limit: usize) {
    crate::mods::set_provider_limit(provider, limit);
}

#[tauri::command]
pub async fn clear_provider_cache(provider: ProviderKind) {
    crate::mods::clear_provider_cache(provider);
}

#[tauri::command]
pub async fn purge_stale_provider_cache(provider: ProviderKind) {
    crate::mods::purge_stale_provider_cache(provider);
}

#[tauri::command]
pub async fn get_extended_mod_info(mod_jar_info: ModJarInfo) -> Result<ExtendedModInfo, String> {
    crate::mods::get_extended_mod_info(mod_jar_info).await
}
